src/userdev/java/net/minecraftforge/gradle/userdev/dependency/DependencyFilter.java

```diff
@@ -0,0 +1,121 @@
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+package net.minecraftforge.gradle.userdev.dependency;
+
+import groovy.lang.Closure;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.artifacts.ExternalModuleDependency;
+import org.gradle.api.artifacts.ResolvedDependency;
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.specs.Spec;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+public interface DependencyFilter {
+    /**
+     * Exclude dependencies that match the provided spec.
+     *
+     * @param spec
+     * @return
+     */
+    DependencyFilter exclude(Spec<? super ArtifactIdentifier> spec);
+
+    /**
+     * Include dependencies that match the provided spec.
+     *
+     * @param spec
+     * @return
+     */
+    DependencyFilter include(Spec<? super ArtifactIdentifier> spec);
+
+    /**
+     * Create a spec that matches the provided project notation on group, name, and version
+     *
+     * @param notation
+     * @return
+     */
+    Spec<? super ArtifactIdentifier> project(Map<String, ?> notation);
+
+    /**
+     * Create a spec that matches the default configuration for the provided project path on group, name, and version
+     *
+     * @param notation
+     * @return
+     */
+    Spec<? super ArtifactIdentifier> project(String notation);
+
+    /**
+     * Create a spec that matches dependencies using the provided notation on group, name, and version
+     *
+     * @param notation
+     * @return
+     */
+    Spec<? super ArtifactIdentifier> dependency(Object notation);
+
+    /**
+     * Create a spec that matches the provided dependency on group, name, and version
+     *
+     * @param dependency
+     * @return
+     */
+    Spec<? super ArtifactIdentifier> dependency(Dependency dependency);
+
+    /**
+     * Create a spec that matches the provided closure
+     *
+     * @param spec
+     * @return
+     */
+    Spec<? super ArtifactIdentifier> dependency(Closure<Boolean> spec);
+
+    boolean isIncluded(ResolvedDependency dependency);
+
+    boolean isIncluded(ExternalModuleDependency dependency);
+
+    boolean isIncluded(ArtifactIdentifier dependency);
+
+    final class ArtifactIdentifier {
```

Comment:
Don't we already have one of these?

Reply:
Yes but not as narrowly scoped, which is the point here.
Since I need to validate both ResolvedDependencies as well as Dependencies the one in artifactural is too widely scoped and requires me to supply information i don't have in every situation.

Hence a small reimplementation with just the fields that are needed / supported.